servlet/src/main/java/io/undertow/servlet/api/DeploymentInfo.java

```diff
@@ -107,6 +107,7 @@
     private boolean escapeErrorMessage = true;
     private boolean sendCustomReasonPhraseOnError = false;
     private boolean useCachedAuthenticationMechanism = true;
+    private boolean preservePathOnForward = true;
```

Comment:
I think this should be `false` by default. Because, if this is set to `true` by default, it violates the servlet spec. Please refer to the Servlet API document [HttpServletRequest#getRequestURL](https://docs.oracle.com/javaee/7/api/javax/servlet/http/HttpServletRequest.html#getRequestURL--) which says:

> If this request has been forwarded using RequestDispatcher.forward(javax.servlet.ServletRequest, javax.servlet.ServletResponse), the server path in the reconstructed URL must reflect the path used to obtain the RequestDispatcher, and not the server path specified by the client.

By this change, the option will be able to be configurable from DeploymentInfo through the servlet extension mechanism. However, I think it's also better to add a new system property as a configurable switch for users who would like to revert the behavior to the previous version of JBoss (JBoss EAP 6.x). Because just adding a system property is easier than creating/adding the servlet extension mechanism to their own deployment.

For example:

```suggestion
    private static final boolean DEFAULT_PRESERVE_PATH_OF_FORWARD = Boolean.getBoolean("io.undertow.servlet.preserve-path-of-forward");

    private boolean preservePathOnForward = DEFAULT_PRESERVE_PATH_OF_FORWARD;
```

Reply:
@msfm good catch! But my interpretation of that Javadoc differs a little bit. The actual path used to obtain the RequestDispatcher is the path received by RequestDispatcherImpl, before it is changed by it. In this case, it may or may not be equal to the path specified by the client. If we don't preserve the path on forward, we are returning a path that is different from the path used to obtain the RequestDispatcher, we are returning a path that is resultant of the action performed by RequestDispatcher. @stuartwdouglas any thoughts?

Reply:
For example, `req.getRequestDispatcher("/forward.jsp").forward(req, res)`. "The path used to obtain the RequestDispatcher" == "/forward.jsp" in this case. Therefore, I think the server path in the reconstructed URL must reflect the path "/forward.jsp" as per the spec. 

Anyway, I think the original behavior does not violate the servlet spec. So, IMO, we should keep the original behavior as default (= `preservePathOnForward` should be `false` by default), and add a configurable switch (= add a system property) for users who want the compatible behavior to the previous version of JBoss.

Reply:
The Servlet spec section 9.4 "The Forward Method" says:
> The path elements of the request object exposed to the target servlet must reflect the
> path used to obtain the RequestDispatcher.
> 
> 
> The only exception to this is if the RequestDispatcher was obtained via the
> getNamedDispatcher method. In this case, the path elements of the request object
> must reflect those of the original request.

It feels like that requirement only applies to the resource the request is forwarded to and doesn't say what should be the behaviour back in original resource.

Reply:
I'm taking the default suggestions made by @msfm to a discussion in the JBEAP jira, where this configuration property was requested in the first place, I think those points are worthy to be considered.

Reply:
Just for a reference, here is the mentioned JBEAP Jira https://issues.jboss.org/browse/JBEAP-16397.